Userland/Libraries/LibWeb/HTML/Canvas/CanvasPath.cpp

```diff
@@ -213,4 +213,154 @@ void CanvasPath::rect(double x, double y, double w, double h)
+
+            // 5.1.3. Otherwise, append radius to normalizedRadii.
+            normalized_radii.append(radius_as_dom_point);
+        }
+
+        // 5.2. If radius is a unrestricted double:
+        if (radius.has<double>()) {
+            auto radius_as_double = radius.get<double>();
+
+            // 5.2.1. If radius is infinite or NaN, then return.
+            if (!isfinite(radius_as_double))
+                return {};
+
+            // 5.2.2. If radius is negative, then throw a RangeError.
+            if (radius_as_double < 0)
+                return WebIDL::SimpleException { WebIDL::SimpleExceptionType::RangeError, "roundRect: Radius can not be negative"sv };
+
+            // 5.2.3. Otherwise append «[ "x" → radius, "y" → radius ]» to normalizedRadii.
+            normalized_radii.append(Geometry::DOMPointInit { radius_as_double, radius_as_double });
+        }
+    }
+
+    // 6. Let upperLeft, upperRight, lowerRight, and lowerLeft be null.
+    Geometry::DOMPointInit upper_left {};
+    Geometry::DOMPointInit upper_right {};
+    Geometry::DOMPointInit lower_right {};
+    Geometry::DOMPointInit lower_left {};
+
+    // 7. If normalizedRadii's size is 4, then set upperLeft to normalizedRadii[0], set upperRight to normalizedRadii[1], set lowerRight to normalizedRadii[2], and set lowerLeft to normalizedRadii[3].
+    if (normalized_radii.size() == 4) {
+        upper_left = normalized_radii.at(0);
+        upper_right = normalized_radii.at(1);
+        lower_right = normalized_radii.at(2);
+        lower_left = normalized_radii.at(3);
+    }
+
+    // 8. If normalizedRadii's size is 3, then set upperLeft to normalizedRadii[0], set upperRight and lowerLeft to normalizedRadii[1], and set lowerRight to normalizedRadii[2].
+    if (normalized_radii.size() == 3) {
+        upper_left = normalized_radii.at(0);
+        upper_right = lower_left = normalized_radii.at(1);
+        lower_right = normalized_radii.at(2);
+    }
+
+    // 9. If normalizedRadii's size is 2, then set upperLeft and lowerRight to normalizedRadii[0] and set upperRight and lowerLeft to normalizedRadii[1].
+    if (normalized_radii.size() == 2) {
+        upper_left = lower_right = normalized_radii.at(0);
+        upper_right = lower_left = normalized_radii.at(1);
+    }
+
+    // 10. If normalizedRadii's size is 1, then set upperLeft, upperRight, lowerRight, and lowerLeft to normalizedRadii[0].
+    if (normalized_radii.size() == 1)
+        upper_left = upper_right = lower_right = lower_left = normalized_radii.at(0);
+
+    // 11. Corner curves must not overlap. Scale all radii to prevent this:
+    // 11.1. Let top be upperLeft["x"] + upperRight["x"].
+    // 11.2. Let right be upperRight["y"] + lowerRight["y"].
+    // 11.3. Let bottom be lowerRight["x"] + lowerLeft["x"].
+    // 11.4. Let left be upperLeft["y"] + lowerLeft["y"].
+    // 11.5. Let scale be the minimum value of the ratios w / top, h / right, w / bottom, h / left.
```

Comment:
I saw at one point you had all these as separate variables. I think that's a bit easier to read, but it's not a big issue.